ccm_web/server/appHandler.ts

```diff
@@ -0,0 +1,146 @@
+  isDev: false
+  staticPath: string
+}
+
+interface WebpackDevOutput {
+  middleware: WebpackDevMiddleware & NextHandleFunction
+  publicPath: string
+}
+
+// TO DO: handle configured host name for prod
+class AppHandler {
+  private readonly config: Config
+  private readonly apiRouter: Router
+  private readonly envOptions: DevOptions | ProdOptions
+
+  constructor (
+    config: Config,
+    envOptions: DevOptions | ProdOptions,
+    apiRouter: Router
+  ) {
+    this.config = config
+    this.envOptions = envOptions
+    this.apiRouter = apiRouter
+  }
+
+  static handleProdClient (app: Express, staticPath: string): void {
+    console.log('Loading dist/client...')
+    app.use(express.static(staticPath))
+
+    app.get('*', (req: Request, res: Response) => {
+      res.sendFile(path.join(staticPath, 'index.html'))
+    })
+  }
+
+  /* TO DO: Investigate some way to dynamically import webpack stuff? */
+  static setupWebpackMiddleware (): WebpackDevOutput {
+    console.log('Setting up webpack-dev-middleware...')
+
+    const publicPath = devWebpackConfig.output?.publicPath
+    if (publicPath === undefined) throw Error('Webpack publicPath was not properly defined!')
+
+    const compiler = webpack(devWebpackConfig)
+    const webpackDevMid = webpackDevMiddleware(compiler, {
+      publicPath: String(publicPath),
+      // Just copies what's used in memory to client/build (changing doesn't do anything)
+      writeToDisk: true
+    })
+    return { middleware: webpackDevMid, publicPath: String(publicPath) }
+  }
+
+  // LTIJS docs: https://cvmcosta.me/ltijs/#/
+  async setupLTI (): Promise<Express> {
+    const { server, db, lti } = this.config
+
+    const dbPlugin = new Database(
+      db.name,
+      db.user,
+      db.password,
+      { host: db.host, dialect: 'postgres', logging: false }
+    )
+
+    let serverAddon
+    let staticPath
+    if (this.envOptions.isDev) {
+      const output = AppHandler.setupWebpackMiddleware()
+      serverAddon = (app: Express) => app.use(output.middleware)
+      staticPath = output.publicPath
+    } else {
+      staticPath = this.envOptions.staticPath
+    }
+
+    const provider = Provider.setup(
+      lti.encryptionKey, // Key used to sign cookies and tokens
+      { plugin: dbPlugin },
+      {
+        // Reserved routes
+        appRoute: '/',
+        loginRoute: '/login',
+        keysetRoute: '/keys',
+        // Set secure to true if the testing platform is in a different domain and https is being used
+        // Set sameSite to 'None' if the testing platform is in a different domain and https is being used
+        cookies: {
+          secure: true,
+          sameSite: 'None'
+        },
+        serverAddon: serverAddon,
+        staticPath: staticPath
+      }
+    )
+
+    provider.app.use('/api', this.apiRouter)
+
+    // Set lti launch callback
+    // When receiving successful LTI launch redirects to app.
+    provider.onConnect((token: any, req: Request, res: Response) => {
```

Comment:
As part of LTI launch you should receive an `idtoken`. is this `token` hold that info. I am trying to console.log this but it dumps out something else. 
I don't see you are extracting the Launch variables like role, course, terms, user info. IS that something you are still working on?

Reply:
Yeah, I know we need to do this, but I wasn't sure the best way to go about it yet. I think some of the information is already stored in the database, or it could be extracted from the JWT sent to the client. Yeah, there's something weird about logging the token here I've seen, I'm not sure exactly what's happening. I'll investigate and report back. We may want to split handling that info out into a separate issue.

Reply:
Yeah.. it is storing it in the different tables. 
`select * From idtokens; select * from accesstokens;select * From contexttokens;`
LTI JS is implementing this storing the idtoken in different ways. It is not in one place but separate tables.

Either you have to access the `idtoken` when request comes to `app.ts` but I am assuming LTIJS is handling it before it is available in request.

Understanding how the access the id_token content is important. Even if it stores in DB we need to store it server cache or session. since we don't want to keep on querying the DB when say merge feature required data than BE needs to aware if he is an instructor or service admin

Reply:
For now, how do you feel about just pulling the custom fields (username, roles, course) you want from the `contexttokens` table and sending them up with `/api/globals`? If we pass them down the component tree, they'll be available. Later we could also look at something like `react-redux`: https://react-redux.js.org/introduction/getting-started

Reply:
I don't yet feel there is a need to redux. This is my thinking...
The Frontend needs to know what is the user role for showing the appropriate content. And based on the request the backend will server up the content. Sending the details of course/term/user I feel could be kept BE. BE need to be aware of user and use the appropriate scoped token to serve up the content.

 With the scoped token approach the token needs to be stored in DB once generated. but should be loaded in from DB and stored in session/cache so that further API call it doesn't need a DB lookup.

Reply:
Mmm, I see. Can we push this off to a separate issue / can you create such an issue and assign it to me? I'd like to do some data modeling for user info, decide what ORM/ database client to use (probably `sequelize` to minimize dependencies) and think more about this caching approach. I don't think processing this information is a required part of getting LTI working (related but not necessarily in scope for the issue).

Reply:
OK, This can be done as a separate issue.

Reply:
w.r.t to Redux or some Client caching. If you look at the mockup we show lot of sections in the courses like merge sections, add_non_um_user, 3rd party gradebook trimmer. We can cache sections info. between view. But practically it would be a rare case that a instructor merges a sections and add non-umich user on the same day to keep that in cache.

We can very well make the API call for each case. but later if we see redundant patterns like this we could improve. 

Some thoughts....

Reply:
Thanks for creating the issues. Definitely need to think more about how much caching will be needed, but that's a later step.